### ISSUE TITLE
[Serve] manifold hyper-connection projection accumulates through float atomicAdd, making mHC decode non-reproducible

### ISSUE THREAD

**Issue by flaviusburca** [invergent-ai/surogate#163]
**Severity:** low &nbsp;·&nbsp; **Location:** [`csrc/src/serve/ops/hyper_connection/manifold.cu:69`](https://github.com/invergent-ai/surogate/blob/7f12387624e04c498dbd8df879eaeff13dea0b25/csrc/src/serve/ops/hyper_connection/manifold.cu#L69) &nbsp;·&nbsp; **Audited at:** `7f123876`

### Defect
`manifold_hyper_connection_mix` splits the projection across `slices` CTAs per token and each CTA `atomicAdd`s its partial sum of squares and its `rows` partial dot products into the same FP32 `partials` slot. Floating-point addition is not associative and CTA completion order is not fixed, so `partials` differs bit-for-bit between runs whenever `slices > 1`. `slices` is chosen from the live SM count and the token count, so a decode round (tokens=1) on a 170-SM card takes `wanted = 340`, clamped by `slice_cap = ceil(width/256)` — 80 slices at GLM-5.3's 4x5120 residual. Those partials feed `rsqrtf`, the `pre`/`post` sigmoids and the Sinkhorn matrix of every layer.

### Failure scenario
GLM-5.3 (45 mHC layers, streams=4, hidden=5120) decoding a single greedy request twice with identical input on the same GPU. Each layer's `partials[token*(rows+1)+slot]` is the sum of 80 atomicAdds whose order differs between the two runs, so `inv = rsqrtf(accumulated[0]/width + rms_eps)` and the 24 projection logits differ in the last few mantissa bits; the Sinkhorn iteration amplifies that through 20 divisions per layer. After 45 layers the logit gap between the top two tokens can be smaller than the accumulated difference, and the two runs emit different tokens from the same prompt with temperature 0 — an output divergence with no input change.

### Evidence
````cpp
manifold.cu:64-71
    __syncthreads();
    if (static_cast<int>(threadIdx.x) <= rows) {
        const int slot = static_cast<int>(threadIdx.x);
        float sum      = 0.0F;
        for (int w = 0; w < kWarps; ++w) { sum += reduce[w * (rows + 1) + slot]; }
        atomicAdd(&partials[static_cast<std::int64_t>(token) * (rows + 1) + slot], sum);
    }

slices chosen at runtime, manifold.cu (host):
        const int wanted = (2 * multiprocessors + tokens - 1) / tokens;
        slices           = std::max(1, std::min(slice_cap, wanted));
    const dim3 project_grid(static_cast<unsigned>(slices), static_cast<unsigned>(tokens));
````

### Suggested fix
Give each slice its own row of the partial buffer (`partials[(token*slices + slice)*(rows+1) + slot]`, zeroed the same way) and have `manifold_reduce_kernel` sum the `slices` rows in a fixed index order. That keeps the split-K parallelism the comment justifies while making the result a deterministic function of the input.

### Verification
Found in an audit of `csrc/src/serve`, then upheld by two independent verifiers: one tried to refute it, the other traced a concrete path to it.
Corrections the verifiers made to the original report:

> manifold_project_kernel (manifold.cu:36-72) launches one CTA per (slice, token), and each one calls atomicAdd on its partial sum into a shared FP32 slot `partials[token*(rows+1)+slot]` (manifold.cu:70). FP32 addition is not associative and nothing fixes the order in which CTAs finish, so whenever slices > 1 the accumulated sum of squares and the 24 projection dots can differ by a few ulps from run to run. slices > 1 always holds on a decode round (tokens=1). For GLM-5.3 (config.h: hidden=4096, 4 streams, width 16384) the round uses slice_cap = 64 slices, not 80. These values flow without any reordering into rsqrtf, the pre/post sigmoids and the Sinkhorn (manifold.cu:92-159) at every one of the 45 layers. The result is bit-level non-reproducibility of the collapse and recombine weights. The Sinkhorn is contractive, so it does not amplify the difference. A different greedy token appears only when the top logits are nearly tied. Fix: write each slice's sums to its own per-(token, slice) slot, with no atomicAdd, and add them in slice order in manifold_reduce_kernel.

> The defect is at manifold.cu:70, not 69. `manifold_project_kernel` splits the projection into (slice, token) CTAs, and each one `atomicAdd`s its FP32 slice sum into a shared `partials[token*(rows+1)+slot]`. Because CTA completion order is not fixed and FP32 addition is not associative, the sum of squares and the 24 projection dots are not bit-reproducible when slices > 1.
> 
> For GLM-5.3-Flash (hidden 4096, 4 streams, width 16384), decode uses slices = min(ceil(16384/256), 2*MP) = 64, not 80. The bit differences reach pre/post/comb and, through BF16 rounding in the collapse and combine, can reach the residual.
> 
> Since `slices` also depends on the token count, the same token's mixing also differs between a batched prefill and a lone decode.
> 
> Greedy-output divergence is possible when the top two logits are nearly tied, but it has not been demonstrated.

_Line numbers quoted by the verifiers may refer to a later tree; the permalinks above are pinned to `7f123876`._

**Comment by flaviusburca** [comment]
Fixed in commit `e825224c8d360458469b0f3def8f908c1fd1b27e` (committed locally; pending push).

Write each projection slice into its own workspace row and sum those rows in a fixed order. Keep the adaptive launch grid, remove the atomic accumulator and zeroing launch, and bound scratch capacity across the full token interval on the selected GPU.

Validation: The old kernel failed repeated eager and captured bit comparisons for decode and prefill. The updated mHC suite passes on GPU 0, including 24 eager and 24 graph repetitions per geometry, one through four streams, maximum residual width, FP64 numerical references, mass conservation and a workspace interval crossing an adaptive split boundary. Compute Sanitizer reports zero errors.
